R/formula.R

```diff
@@ -0,0 +1,422 @@
+
+        combined_terms <- c()
+        for (v1 in split_terms[[1]])
+          for (v2 in split_terms[[2]])
+            combined_terms <- c(combined_terms, paste0(v1, ":", v2))
+
+        formula <- update(formula, paste0(". ~ . - ", current,
+                                          " + ",
+                                          paste(combined_terms, collapse=" + ")))
+        ## no need to update the data because the interaction terms
+        ## do not appear as features
+      }
+    }
+  }
+
+  tryCatch(model.matrix(formula, data=new_data),
+           error = function (e) print(e))
+  list(formula=formula, data=new_data)
+}
+
+split_linear_term <- function(term, term_formula, data) {
+  ## extract individual variables
+  ## simple_mm <- model.matrix(term_formula, data=data)
+  ## assign <- attr(simple_mm, "assign")
+  ## appearances <- assign[assign != 0]
+
+  appearances <- ncol(data[, term])
+
+  if (appearances > 1)
+    split_term <- sapply(1:appearances,
+                         function(i) paste0(term, i))
+  else
+    split_term <- term
+  split_term
+}
+
+count_terms_in_submodel <- function(submodel) {
+  if (class(submodel) != "formula")
+    submodel <- as.formula(paste0("~ ", submodel))
+  tt <- extract_terms_response(submodel)
+  ind_int_terms <- length(tt$individual_terms) + length(tt$int_terms)
+  group_terms <- sum(unlist(lapply(tt$group_terms, count_terms_in_group_term)))
+  if (is.null(group_terms))
+    group_terms <- 0
+  ind_int_terms + group_terms
+}
+
+count_terms_in_group_term <- function(term) {
+  term <- gsub("[\\(\\)]", "", flatten_group_terms(term))
+
+  chunks <- strsplit(term, " \\| ")[[1]]
+  variables <- strsplit(chunks[1], " \\+ ")[[1]]
+
+  ## split possible interaction terms
+  int_t <- str_match(variables, ":")
+  int_v <- variables[!is.na(int_t)]
+  lin_v <- variables[is.na(int_t)]
+
+  if ("0" %in% variables) {
+    group_intercept <- FALSE
+    variables <- setdiff(variables, "0")
+  } else {
+    group_intercept <- TRUE
+    if (!("1" %in% variables))
+      variables <- c(variables, "1")
+  }
+
+  length(variables)
+}
+
+sort_submodels_by_size <- function(submodels) {
+  size <- lapply(submodels, count_terms_in_submodel)
+  df <- as.data.frame(cbind(submodels=submodels, size=unlist(size)))
+  df$size <- as.numeric(df$size)
+  ordered <- df[order(df$size), ]
+
+  groups <- list()
+  for (size in unique(ordered$size))
+    groups[[size]] <- as.character(ordered$submodels[ordered$size == size])
+
+  ord_list <- groups
+  ord_list_nona <- lapply(ord_list, function(l) l[!is.na(l)])
+  ord_list_nona[!is.na(ord_list_nona)]
+}
+
+count_variables_chosen <- function(refmodel, list_of_terms) {
+  if (length(list_of_terms) == 0)
+    return(0)
+  form <- subset_formula(refmodel$formula, unique(unlist(list_of_terms)),
+                         data=refmodel$fetch_data())$formula
+  count_terms_in_submodel(flatten_formula(form))
+}
+
+is.redundant <- function(refmodel, current, new) {
```

Comment:
renamed twice. We are checking if the submodel after adding `new` predictor is redundant with the current submodel. This can happen as when we sequentially increase the models some of the predictors may be included in a previous step.